lib/document.js

```diff
@@ -0,0 +1,1469 @@
+        data: {
+          first_published_at: $first_published_at, created_by_email: $created_by_email, headline: $headline, locale_code: $locale_code, published: $published, content: $content, custom_byline: $custom_byline, facebook_description: $facebook_description, facebook_title: $facebook_title, search_description: $search_description, search_title: $search_title, twitter_description: $twitter_description, twitter_title: $twitter_title, main_image: $main_image
+        }
+      }, 
+      article_sources: {
+        data: $article_sources,
+        on_conflict: {constraint: article_source_article_id_source_id_key, update_columns: article_id}
+      },
+      category_id: $category_id, 
+      id: $id, 
+      slug: $slug, 
+      canonical_url: $canonical_url,
+      article_google_documents: {
+        data: {
+          google_document: {
+            data: {
+              document_id: $document_id, locale_code: $locale_code, url: $url
+            }, 
+            on_conflict: {
+              constraint: google_documents_organization_id_document_id_key, update_columns: locale_code
+            }
+          }
+        }, 
+        on_conflict: {
+          constraint: article_google_documents_article_id_google_document_id_key, update_columns: google_document_id
+        }
+      }
+    }, 
+    on_conflict: {
+      constraint: articles_pkey, update_columns: [category_id, canonical_url, slug, updated_at]
+    }
+  ) {
+    returning {
+      id
+      canonical_url
+      slug
+      updated_at
+      created_at
+      article_google_documents {
+        id
+        google_document {
+          document_id
+          locale_code
+          locale {
+            code
+            name
+          }
+          url
+          id
+        }
+      }
+      article_sources {
+        source {
+          affiliation
+          age
+          email
+          ethnicity
+          gender
+          id
+          name
+          phone
+          race
+          role
+          sexual_orientation
+          zip
+        }
+      }
+      category {
+        slug
+      }
+      article_translations(where: { locale_code: {_eq: $locale_code}}, order_by: {id: desc}, limit: 1) {
+        id
+        article_id
+        locale_code
+        published
+        headline
+        first_published_at
+        last_published_at
+      }
+      published_article_translations(where: {locale_code: {_eq: $locale_code}}) {
+        article_translation {
+          id
+          first_published_at
+          last_published_at
+          locale_code
+        }
+      }
+    }
+  }
+}`;
+
+export async function saveArticle(params) {
+  let returnValue = {
+    status: 'success',
+    message: '',
+    data: {},
+  };
+
+  let data = params['data'];
+  // console.log('data keys:', Object.keys(data).sort());
+  let articleAuthors = data.article_authors;
+  let articleTags = data.article_tags;
+  delete data.article_authors;
+  delete data.article_tags;
+
```

Comment:
Because further down in the function `data` is passed directly as the variables for a graphql mutation. Article authors and tags are handled separately. If these two attributes remained on `data`, an error would be thrown by Hasura about unexpected variables.

Reply:
Ah, cool.